### PR TITLE
8332153: RISC-V: enable tests and add comment for vector shift instruct (shared by vectorization and Vector API)

### PR DESCRIPTION
**Pull request by Hamlin-Li**
Hi,
Can you help to review this patch?
For vector shift instruct, some corresponding tests are not enabled, this is to enable them.
And the way how vector shift instruct works is not clear, especially both vectorization (SLP in jdk) and Vector API share the same instruct's in riscv_v.ad, so also added some comment to clarify it.

Thanks

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8332153](https://bugs.openjdk.org/browse/JDK-8332153): RISC-V: enable tests and add comment for vector shift instruct (shared by vectorization and Vector API) (**Enhancement** - P4)

### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19265/head:pull/19265` \
`$ git checkout pull/19265`

Update a local copy of the PR: \
`$ git checkout pull/19265` \
`$ git pull https://git.openjdk.org/jdk.git pull/19265/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19265`

View PR using the GUI difftool: \
`$ git pr show -t 19265`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19265.diff">https://git.openjdk.org/jdk/pull/19265.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19265#issuecomment-2114951877)

### CODE REVIEW

File: src/hotspot/cpu/riscv/riscv_v.ad [hunk 1]
```diff
@@ -1782,6 +1782,32 @@ instruct replicateD(vReg dst, fRegD src) %{
...
+//  2. similarly, for long "with the mask value 0x3f (0b111111)"
+// check https://docs.oracle.com/javase/specs/jls/se21/html/jls-15.html#jls-15.19 for details.
+//
+// Shift behaviour in Vector APi is defined as:
+//   e.g. for ASHR, "a>>(n&(ESIZE*8-1))"
+//   this behaviour is the same as shift instrunction's in riscv vector extension.
+// check https://docs.oracle.com/en/java/javase/21/docs/api/jdk.incubator.vector/jdk/incubator/vector/VectorOperators.html#ASHR
+// and https://github.com/riscv/riscv-v-spec/blob/master/v-spec.adoc#116-vector-single-width-shift-instructions for details.
+//
+// Although the difference between these 2 behaviours, the same shift instruct's of byte and short are
```
[the remark this answers is not in the document]
Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/riscv_v.ad]: Thanks, fixed.

File: src/hotspot/cpu/riscv/riscv_v.ad [hunk 2]
```diff
@@ -1782,6 +1782,32 @@ instruct replicateD(vReg dst, fRegD src) %{
...
+//
+// Shift behaviour in vectorization is defined by java language spec, which includes:
+//  1. "If the promoted type of the left-hand operand is int, then only the five lowest-order bits of
+//      the right-hand operand are used as the shift distance. It is as if the right-hand operand were
+//      subjected to a bitwise logical AND operator & (§15.22.1) with the mask value 0x1f (0b11111).
+//      The shift distance actually used is therefore always in the range 0 to 31, inclusive."
+//  2. similarly, for long "with the mask value 0x3f (0b111111)"
+// check https://docs.oracle.com/javase/specs/jls/se21/html/jls-15.html#jls-15.19 for details.
+//
+// Shift behaviour in Vector APi is defined as:
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/riscv_v.ad:1796]: Nit: s/APi/API/

File: test/hotspot/jtreg/compiler/vectorization/runner/ArrayShiftOpTest.java [hunk 3]
```diff
@@ -99,6 +99,9 @@ public long[] longCombinedRotateShift() {
     @Test
     @IR(applyIfCPUFeatureOr = {"asimd", "true", "sse2", "true"},
         counts = {IRNode.RSHIFT_VI, ">0"})
+    @IR(applyIfPlatform = {"riscv64", "true"},
+        applyIfCPUFeature = {" v ", "true"},
```
Comment (RealFYang) [inline test/hotspot/jtreg/compiler/vectorization/runner/ArrayShiftOpTest.java:103]: Is the spaces around `v` necessary?

File: test/hotspot/jtreg/compiler/c2/aarch64/TestVectorShiftShorts.java [hunk 4]
```diff
@@ -27,7 +27,8 @@
  * @bug 8255287
  * @summary aarch64: fix SVE patterns for vector shift count
  *
- * @requires os.arch == "aarch64" & vm.compiler2.enabled
+ * @requires vm.compiler2.enabled
+ * @requires os.arch == "aarch64" | os.arch == "riscv64"
```
Comment (RealFYang) [inline test/hotspot/jtreg/compiler/c2/aarch64/TestVectorShiftShorts.java:31]: Seems a bit weird to enable tests under `test/hotspot/jtreg/compiler/c2/aarch64/` for other cpus?

File: test/hotspot/jtreg/compiler/vectorization/runner/ArrayShiftOpTest.java (hunk 3 above)
Comment (Hamlin-Li) [reply test/hotspot/jtreg/compiler/vectorization/runner/ArrayShiftOpTest.java]: Thanks for catching, fixed.

File: src/hotspot/cpu/riscv/riscv_v.ad (hunk 2 above)
Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/riscv_v.ad]: fixed.

File: test/hotspot/jtreg/compiler/c2/aarch64/TestVectorShiftShorts.java (hunk 4 above)
Comment (Hamlin-Li) [reply test/hotspot/jtreg/compiler/c2/aarch64/TestVectorShiftShorts.java]: Yeh, it's bit, removed ` | os.arch == "riscv64"`.

I can also modify the summary and enable the test for riscv if needed.

File: test/hotspot/jtreg/compiler/c2/aarch64/TestVectorShiftShorts.java [hunk 5]
```diff
@@ -27,7 +27,8 @@
  * @bug 8255287
  * @summary aarch64: fix SVE patterns for vector shift count
  *
- * @requires os.arch == "aarch64" & vm.compiler2.enabled
+ * @requires vm.compiler2.enabled
+ * @requires os.arch == "aarch64"
```
[the remark this answers is not in the document]
Comment (Hamlin-Li) [reply test/hotspot/jtreg/compiler/c2/aarch64/TestVectorShiftShorts.java]: No, fixed.

**Review by RealFYang — APPROVED** [review]
LGTM assuming these tests pass with RVV. Thanks.

**Comment by Hamlin-Li** [comment]
Yes, the tests passed.

Thanks @RealFYang @luhenry for your reviewing.

/integrate

**Pull request closed without merging**